client/boot/index.js

```diff
@@ -161,8 +160,39 @@ function boot() {
 
 	translatorJumpstart.init();
 
-	reduxStore = createReduxStore();
+	localforage.config( {
+		name: 'calypso',
+		storeName: 'calypso_store',
+		description: 'Calypso Storage'
+	} );
+	localforage.getItem( 'redux-state', ( error, initialState ) => {
+		let reduxStore;
+		if ( error ) {
+			debug( 'failed to load initial redux-store state', error );
+			reduxStore = createReduxStore();
+		} else {
+			debug( 'loading redux-store with initial state', initialState );
+			reduxStore = createReduxStore( initialState || {} );
+		}
+
+		reduxStore.subscribe( function() {
+			let state = Object.assign( {}, reduxStore.getState() );
+			//remove dirty parts of the redux tree
+			delete state.sites;
+			delete state.themes;
+			localforage.setItem( 'redux-state', state, ( setError ) => {
+				if ( setError ) {
+					debug( 'failed to set redux-store state', setError );
+				}
+			} );
+		} );
+
+		reduxStoreReady( reduxStore );
+	} );
```

Comment:
Considering how central this loader will be I think it would be really helpful to declare the handler function outside of the call to `getItem()`. _**We should definitely have tests to handle bizarre cases here.**_

On that point, I think we could also benefit from using the Promise interface to `localforage` because it would allow us to build up more robust error-handling without mixing the logic with our successful data handling.

```js
local forage.getItem()
	.then( loadInitialState )
	.catch( handleStateLoadError );
```